File: app/src/main/java/com/wuguangxin/simple/adapter/GameAdapter.java
package com.wuguangxin.simple.adapter;

import android.content.Context;
import android.view.View;

import com.wuguangxin.adapter.BaseRecyclerAdapter;
import com.wuguangxin.adapter.BaseViewHolder;
import com.wuguangxin.simple.R;
import com.wuguangxin.simple.bean.GameDataBean;
import com.wuguangxin.simple.view.SquareCheckBox;

import java.util.LinkedList;
import java.util.List;

public class GameAdapter extends BaseRecyclerAdapter<GameDataBean, BaseViewHolder> {

    private LinkedList<GameDataBean> resultData = new LinkedList<>();
    private int grayColor;
    boolean full = false;

    public GameAdapter(Context context, List<GameDataBean> list) {
        super(context, list);
        grayColor = context.getResources().getColor(R.color.gray);
    }

    @Override
    public int getLayoutId() {
        return R.layout.item_game;
    }

    @Override
    public BaseViewHolder createViewHolder(View view, int position) {
        return new BaseViewHolder(view);
    }

    public List<GameDataBean> getResultData() {
        return resultData;
    }

    @Override
    public void setList(List<GameDataBean> list) {
        full = false;
        resultData.clear();
        resultData = new LinkedList<>();
        super.setList(list);
    }

    @Override
    public void bindViewData(BaseViewHolder viewHolder, GameDataBean gameDataBean, int position, int type) {
        viewHolder.setIsRecyclable(false); // false: view不重复利用，降低性能
        SquareCheckBox checkBox = (SquareCheckBox) viewHolder.getItemView();
        if (gameDataBean.getChecked()) {
            checkBox.setBackgroundColor(gameDataBean.getColor());
            checkBox.setOnCheckedChangeListener(null);
        } else if (!gameDataBean.getEnable()) {
            checkBox.setBackgroundColor(grayColor);
            checkBox.setOnCheckedChangeListener(null);
        } else {
            checkBox.setOnCheckedChangeListener((compoundButton, checked) -> {
                if (checked) {
                    if (full) {
                        return;
                    }
                    resultData.add(gameDataBean);
                    gameDataBean.setChecked(true);
                    gameDataBean.setEnable(false);
                    checkBox.setBackgroundColor(gameDataBean.getColor());
                    checkBox.setEnabled(gameDataBean.getEnable());
                    if (onStatusListener != null) {
                        onStatusListener.onChecked(gameDataBean);
                    }
                    if (resultData.size() >= 12) {
                        int itemCount = getItemCount();
                        for (int i = 0; i < itemCount; i++) {
                            getItem(i).setEnable(false);
                        }
                        full = true;
                        notifyDataSetChanged();
                    }
                }
            });
        }

        checkBox.setChecked(gameDataBean.getChecked());
        checkBox.setEnabled(gameDataBean.getEnable() || resultData.size() >= 12);
        checkBox.setText(gameDataBean.getText());
    }

    private OnStatusListener onStatusListener;

    public void setOnStatusListener(OnStatusListener listener) {
        this.onStatusListener = listener;
    }

    public interface OnStatusListener {
        void onChecked(GameDataBean gameDataBean);
    }


}

File: app/src/main/java/com/wuguangxin/simple/adapter/GameAdapter.kt
package com.wuguangxin.simple.adapter

import com.wuguangxin.simple.bean.GameDataBean
import com.wuguangxin.simple.R
import com.wuguangxin.simple.view.SquareCheckBox
import android.widget.CompoundButton
import com.chad.library.adapter.base.BaseQuickAdapter
import com.chad.library.adapter.base.viewholder.BaseViewHolder
import java.util.*

class GameAdapter(list: MutableList<GameDataBean>) :
    BaseQuickAdapter<GameDataBean, BaseViewHolder>(R.layout.item_game, list) {
    private var resultData = LinkedList<GameDataBean>()
    var full = false

    fun reset() {
        full = false
        resultData.clear()
    }

    fun getResultData(): List<GameDataBean> {
        return resultData
    }

    override fun convert(holder: BaseViewHolder, item: GameDataBean) {
        holder.setIsRecyclable(false) // false: view不重复利用，降低性能
        val checkBox = holder.itemView as SquareCheckBox
        if (item.checked) {
            checkBox.setBackgroundColor(item.color)
            checkBox.setOnCheckedChangeListener(null)
        } else if (!item.enable) {
            checkBox.setBackgroundResource(R.color.gray)
            checkBox.setOnCheckedChangeListener(null)
        } else {
            checkBox.setOnCheckedChangeListener { _: CompoundButton?, checked: Boolean ->
                if (checked) {
                    if (full) {
                        return@setOnCheckedChangeListener
                    }
                    resultData.add(item)
                    item.checked = true
                    item.enable = false
                    checkBox.setBackgroundColor(item.color)
                    checkBox.isEnabled = item.enable
                    onStatusListener?.onChecked(item)
                    if (resultData.size >= 12) {
                        resultData.forEach {
                            it.enable = false
                        }
                        full = true
                        notifyItemRangeChanged(0, itemCount)
                    }
                }
            }
        }
        checkBox.isChecked = item.checked
        checkBox.isEnabled = item.enable || resultData.size >= 12
        checkBox.text = item.text
    }

    private var onStatusListener: OnStatusListener? = null
    fun setOnStatusListener(listener: OnStatusListener?) {
        onStatusListener = listener
    }

    interface OnStatusListener {
        fun onChecked(gameDataBean: GameDataBean?)
    }
}